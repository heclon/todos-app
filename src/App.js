import React, { Component } from "react";
import { Container, Row, Col, Toast, ToastBody, ToastHeader } from "reactstrap";
import ModalAddEdit from "./Components/Modals/ModalAddEdit";
import DataTable from "./Components/Tables/DataTable";

function countByPriority(items) {
let counterHighPriority = 0;
let counterMediumPriority = 0;
let counterLowPriority = 0;
  for (const item of items) {
    if (item.priorityValue === 1) counterHighPriority++;
    if (item.priorityValue === 2) counterMediumPriority++;
    if (item.priorityValue === 3) counterLowPriority++;
  }
  
return [ counterHighPriority, counterMediumPriority, counterLowPriority ]
}
class App extends Component {
  state = {
    items: [],
    highPriority: 0,
    mediumPriority: 0,
    lowPriority: 0,
  };

  getItems() {
    this.setState({ items: [] });
  }

  addItemToState = (item) => {
    this.setState((prevState) => ({
      items: [...prevState.items, item],
    }));
    this.updateCounters()
  };

  updateState = (item) => {
    const itemIndex = this.state.items.findIndex((data) => data.id === item.id);
    const newArray = [
      ...this.state.items.slice(0, itemIndex),
      item,
      ...this.state.items.slice(itemIndex + 1),
    ];
    this.setState({ items: newArray });
    this.updateCounters()
  };

  deleteItemFromState = (id) => {
    const updatedItems = this.state.items.filter((item) => item.id !== id);
    this.setState({ items: updatedItems });
    this.updateCounters(this.state.items)
  };

  componentDidMount() {
    this.getItems();
  }


  
  updateCounters(){
    const items = this.state.items
    const [ counterHighPriority, 
      counterMediumPriority, 
      counterLowPriority ] =  countByPriority(items)

      
    this.setState({
      highPriority: counterHighPriority,
      mediumPriority: counterMediumPriority,
      lowPriority: counterLowPriority
    })

  }

  render() {

    const [ counterHighPriority, 
      counterMediumPriority, 
      counterLowPriority ] =  countByPriority(this.state.items)

    return (
      <Container className="App">
        <Row>
          <Col>
            <h1 style={{ margin: "20px 0" }}>My To-do's</h1>
          </Col>
        </Row>
        <Row style={{ marginBottom: "3%" }}>
          <Col>
            <ModalAddEdit
              buttonLabel="Add To-do task"
              addItemToState={this.addItemToState}
            />
          </Col>
          <Col>
            <div className="p-3 bg-danger rounded">
              <Toast>
                <ToastHeader>
                  High Priority
                </ToastHeader>
                <ToastBody>
                  {counterHighPriority}
                </ToastBody>
              </Toast>
            </div>
          </Col>
          <Col>
            <div className="p-3 bg-warning rounded">
              <Toast>
                <ToastHeader>
                Medium Priority 
                </ToastHeader>
                <ToastBody>
                {counterMediumPriority}
                </ToastBody>
              </Toast>
            </div>
          </Col>
          <Col>
            <div className="p-3 bg-primary rounded">
              <Toast>
                <ToastHeader>
                  Low Priority
                </ToastHeader>
                <ToastBody>
                {counterLowPriority}
                </ToastBody>
              </Toast>
            </div>
          </Col>
        </Row>
        <Row>
          <Col>
            <DataTable
              items={this.state.items}
              updateState={this.updateState}
              deleteItemFromState={this.deleteItemFromState}
            />
          </Col>
        </Row>
        <Row>
          <Col>
          Created by Hector Longarte</Col>
        </Row>
        <Row>
          <Col>
          Linkedin: </Col>
        </Row>
      </Container>
    );
  }
}

export default App;
