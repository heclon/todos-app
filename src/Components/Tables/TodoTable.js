import React, { Component } from "react";
import { Table } from "reactstrap";
import ModalAddEdit from "../Modals/ModalAddEdit";
import ModalConfirmDelete from "../Modals/ModalConfirmDelete";
class TodoTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortOrder: "INIT",
      items: this.props.items,
    };
  }

  sortByName = () => {
    // console.log(this.state.sortOrder);
    let sortedItems = [];
    if (this.state.sortOrder !== "desc") {
      this.setState({
        sortOrder: "desc",
      });

      sortedItems = this.props.items.sort(function (x, y) {
        let a = x.taskName.toUpperCase(),
          b = y.taskName.toUpperCase();
        return a === b ? 0 : a > b ? 1 : -1;
      });
    } else if (this.state.sortOrder === "desc") {
      this.setState({
        sortOrder: "asc",
      });

      sortedItems = this.props.items.sort(function (x, y) {
        let a = x.taskName.toUpperCase(),
          b = y.taskName.toUpperCase();
        return a === b ? 0 : a > b ? -1 : 1;
      });
    }
    this.setState({
      items: sortedItems,
    });
    // console.log(sortedItems);
  };

  sortByPriority = () => {
    // console.log(this.state.sortOrder);
    let sortedItems = [];
    if (this.state.sortOrder !== "desc") {
      this.setState({
        sortOrder: "desc",
      });

      sortedItems = this.props.items.sort(function (x, y) {
        let a = x.priorityValue,
          b = y.priorityValue;
        return a === b ? 0 : a > b ? 1 : -1;
      });
    } else if (this.state.sortOrder === "desc") {
      this.setState({
        sortOrder: "asc",
      });

      sortedItems = this.props.items.sort(function (x, y) {
        let a = x.priorityValue,
          b = y.priorityValue;
        return a === b ? 0 : a > b ? -1 : 1;
      });
    }
    this.setState({
      items: sortedItems,
    });
    // console.log(sortedItems);
  };

  render() {
    const items = this.props.items.map((item) => {
      return (
        <tr key={item.id}>
          <th scope="row">{item.id}</th>
          <td>{item.taskName}</td>
          <td>{item.priority}</td>
          <td>
            <div style={{ width: "110px" }}>
              <ModalAddEdit
                buttonLabel="Edit To-do"
                item={item}
                updateState={this.props.updateState}
              />{" "}
              <ModalConfirmDelete
                buttonLabel="Delete To-do"
                item={item}
                deleteItemFromState={this.props.deleteItemFromState}
              ></ModalConfirmDelete>
            </div>
          </td>
        </tr>
      );
    });

    return (
      <Table dark responsive hover>
        <thead>
          <tr>
            <th>ID</th>
            <th
              id="task-header"
              style={{ cursor: "pointer" }}
              onClick={() => this.sortByName()}
            >
              Task to-do name
            </th>
            <th
              id="priority-header"
              style={{ cursor: "pointer" }}
              onClick={() => this.sortByPriority()}
            >
              Priority
            </th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>{items}</tbody>
      </Table>
    );
  }
}

export default TodoTable;
