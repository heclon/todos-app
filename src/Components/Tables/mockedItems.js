const mockedItems = [
  {
    id: "hook",
    taskName: "Demo React Hooks",
    priority: "High",
    priorityValue: 1,
  },
  {
    id: "awsx",
    taskName: "AWS Architecture exams",
    priority: "Medium",
    priorityValue: 2,
  },
  {
    id: "awsd",
    taskName: "AWS Development exam",
    priority: "Medium",
    priorityValue: 2,
  },
  {
    id: "pyad",
    taskName: "Python advanced learning",
    priority: "Low",
    priorityValue: 3,
  },
  {
    id: "zing",
    taskName: "Zing-Zing last",
    priority: "Medium",
    priorityValue: 2,
  },
];

export default mockedItems;
