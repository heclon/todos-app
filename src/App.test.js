import React from "react";
import { mount } from "enzyme";
import App from "./App";
import TodoTable from "./Components/Tables/TodoTable";
import ModalAddEdit from "./Components/Modals/ModalAddEdit";
import renderer from "react-test-renderer";
import mockedItems from "./Components/Tables/mockedItems";

describe("App tests", () => {
  let appComponent;
  let todoTableComponent;
  beforeEach(() => {
    appComponent = mount(<App />);
    todoTableComponent = mount(
      <TodoTable
        items={mockedItems}
        addItemToState={App.addItemToState}
        updateState={App.updateState}
      />
    );
  });

  describe("App rendering tests", () => {
    it("renders App without crashing", () => {
      renderer.create(<App />);
    });

    it("snapshot test render correctly App component", () => {
      const app = renderer.create(<App />).toJSON();
      expect(app).toMatchSnapshot();
    });

    it("renders the ModalAddEdit to add a new to-do ", () => {
      const modalAddEdit = appComponent.find(ModalAddEdit);
      expect(modalAddEdit).toHaveLength(1);
      expect(modalAddEdit.text()).toBe("Add To-do task");
    });

    it("renders the priority counters correctly", () => {
      const toastsHeaders = appComponent.find(".toast-header");
      expect(toastsHeaders).toHaveLength(3);

      // high priority counter header
      expect(toastsHeaders.at(0).text()).toBe("High Priority");

      // medium priority counter header
      expect(toastsHeaders.at(1).text()).toBe("Medium Priority");

      // low priority counter header
      expect(toastsHeaders.at(2).text()).toBe("Low Priority");

      // and counters should be zero initially
      const toastsValues = appComponent.find(".toast-body");
      expect(toastsValues).toHaveLength(3);

      expect(toastsValues.at(0).text()).toBe("0");
      expect(toastsValues.at(1).text()).toBe("0");
      expect(toastsValues.at(2).text()).toBe("0");
    });
  });

  describe("Priority counters tests", () => {
    it("App displays values on priority counters correctly", () => {
      // set state values
      appComponent.setState({
        items: mockedItems,
        highPriority: 1,
        mediumPriority: 3,
        lowPriority: 1,
      });

      const table = todoTableComponent.find("table");
      const body = table.find("tbody");
      const rows = body.find("tr");
      expect(rows).toHaveLength(5);

      // check for counters to have values 1,3,1
      const toastsValues = appComponent.find(".toast-body");
      expect(toastsValues).toHaveLength(3);

      expect(toastsValues.at(0).text()).toBe("1");
      expect(toastsValues.at(1).text()).toBe("3");
      expect(toastsValues.at(2).text()).toBe("1");
    });

    it("App updates values on priority counters correctly ", () => {
      // set state values
      appComponent.setState({
        items: mockedItems,
        highPriority: 1,
        mediumPriority: 3,
        lowPriority: 1,
      });

      const table = todoTableComponent.find("table");
      const body = table.find("tbody");
      const rows = body.find("tr");
      expect(rows).toHaveLength(5);

      // check for counters to have values 1,3,1
      let toastsValues = appComponent.find(".toast-body");
      expect(toastsValues).toHaveLength(3);

      expect(toastsValues.at(0).text()).toBe("1");
      expect(toastsValues.at(1).text()).toBe("3");
      expect(toastsValues.at(2).text()).toBe("1");

      // change those values and check against new expected values
      const modifiedItem = {
        id: "awsx",
        taskName: "AWS Architecture exams",
        priority: "High",
        priorityValue: 1,
      };
      mockedItems[1] = modifiedItem;

      appComponent.setState({
        items: mockedItems,
        highPriority: 2,
        mediumPriority: 2,
        lowPriority: 1,
      });

      toastsValues = appComponent.find(".toast-body");
      expect(toastsValues).toHaveLength(3);

      expect(toastsValues.at(0).text()).toBe("2");
      expect(toastsValues.at(1).text()).toBe("2");
      expect(toastsValues.at(2).text()).toBe("1");
    });
  });
});
