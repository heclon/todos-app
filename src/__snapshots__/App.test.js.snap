// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests App rendering tests snapshot test render correctly App component 1`] = `
<div
  className="App container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h1
        style={
          Object {
            "margin": "20px 0",
          }
        }
      >
        My To-do's
      </h1>
    </div>
  </div>
  <div
    className="row"
    style={
      Object {
        "marginBottom": "3%",
      }
    }
  >
    <div
      className="col"
    >
      <div>
        <button
          aria-label={null}
          className="btn btn-success"
          onClick={[Function]}
          style={
            Object {
              "float": "left",
              "marginRight": "10px",
            }
          }
          type="button"
        >
          Add To-do task
        </button>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="p-3 bg-danger rounded"
      >
        <div
          className="toast fade"
          role="alert"
        >
          <div
            className="toast-header"
          >
            <strong
              className="mr-auto"
            >
              High Priority
            </strong>
          </div>
          <div
            className="toast-body"
            style={
              Object {
                "fontSize": "20px",
              }
            }
          >
            0
          </div>
        </div>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="p-3 bg-warning rounded"
      >
        <div
          className="toast fade"
          role="alert"
        >
          <div
            className="toast-header"
          >
            <strong
              className="mr-auto"
            >
              Medium Priority
            </strong>
          </div>
          <div
            className="toast-body"
            style={
              Object {
                "fontSize": "20px",
              }
            }
          >
            0
          </div>
        </div>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="p-3 bg-primary rounded"
      >
        <div
          className="toast fade"
          role="alert"
        >
          <div
            className="toast-header"
          >
            <strong
              className="mr-auto"
            >
              Low Priority
            </strong>
          </div>
          <div
            className="toast-body"
            style={
              Object {
                "fontSize": "20px",
              }
            }
          >
            0
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="table-responsive"
      >
        <table
          className="table table-dark table-hover"
        >
          <thead>
            <tr>
              <th>
                ID
              </th>
              <th
                id="task-header"
                onClick={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              >
                Task to-do name
              </th>
              <th
                id="priority-header"
                onClick={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              >
                Priority
              </th>
              <th>
                Actions
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <a
        href="https://linkedin.com/in/hector-longarte"
        rel="noopener noreferrer"
        target="_blank"
      >
        Hector Longarte
      </a>
    </div>
  </div>
</div>
`;
